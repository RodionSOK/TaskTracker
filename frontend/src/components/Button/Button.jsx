import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { Link } from "react-router-dom";
import "./Button.pcss";

const Button = ({
    className,
    children,
    size,
    theme,
    disabled,
    inactive,
    type = "button", // Добавлено значение по умолчанию
    wide,
    href,
    onClick,
    style,
    loading = false,
}) => {
    const componentClassName = classNames("button", className, {
        button_medium: size === "medium",
        button_small: size === "small",
        button_gray: theme === "gray",
        button_black: theme === "black",
        button_red: theme === "red",
        button_wide: wide,
        button_inactive: inactive,
    });

    if (href) {
        return (
            <Link className={componentClassName} to={href}>
                {children}
            </Link>
        );
    }

    const handleClick = (e) => {
        // Для кнопок типа "submit" НЕ вызываем preventDefault()
        if (type !== "submit") {
            e.preventDefault();
        }
        
        if (typeof onClick === "function" && !loading && !disabled) {
            onClick(e);
        }
    };

    return (
        <button
            className={classNames(componentClassName, { "button-loading": loading })}
            disabled={disabled || loading}
            type={type}
            style={style}
            onClick={handleClick} // Исправленный обработчик
        >
            {loading && <div className="loading"></div>}
            <div className={loading ? "children-container" : undefined}>
                {children}
            </div>
        </button>
    );
};

Button.propTypes = {
    className: PropTypes.string,
    children: PropTypes.node,
    size: PropTypes.oneOf(["medium", "small"]),
    theme: PropTypes.oneOf(["gray", "black", "red"]),
    disabled: PropTypes.bool,
    inactive: PropTypes.bool,
    type: PropTypes.oneOf(["button", "submit", "reset"]), // Обновленные типы
    wide: PropTypes.bool,
    href: PropTypes.string,
    onClick: PropTypes.func,
    style: PropTypes.object,
    loading: PropTypes.bool,
};

export default Button;