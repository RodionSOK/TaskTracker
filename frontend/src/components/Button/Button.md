B1.50 Синяя (основная) текстовая кнопка

```jsx
<div className="styleguide__buttons">
    <Button>Текстовая кнопка</Button>
    <Button disabled>Текстовая кнопка</Button>
</div>
```

B1.40 Синяя (основная) текстовая кнопка

```jsx
<div className="styleguide__buttons">
    <Button size="small">Далее</Button>
    <Button size="small" disabled>
        Далее
    </Button>
</div>
```

B2.50 Серая (вторичная) текстовая кнопка

```jsx
<div className="styleguide__buttons">
    <Button theme="gray">Текстовая кнопка</Button>
    <Button theme="gray" disabled>
        Текстовая кнопка
    </Button>
</div>
```

B2.40 Серая (вторичная) текстовая кнопка

```jsx
<div className="styleguide__buttons">
    <Button size="small" theme="gray">
        Далее
    </Button>
    <Button size="small" theme="gray" disabled>
        Далее
    </Button>
</div>
```

Неактивная кнопка (имеет стили отключённой), на которую можно повесить событие

```jsx
<div className="styleguide__buttons">
    <Button inactive onClick={() => alert("click work")}>
        Неактивная кнопка
    </Button>
    <Button disabled onClick={() => alert("click doesn't work")}>
        Отключённая кнопка
    </Button>
</div>
```

Кнопка на всю ширину блока

```jsx
<div className="styleguide__buttons">
    <Button wide>Широкая кнопка</Button>
</div>
```
