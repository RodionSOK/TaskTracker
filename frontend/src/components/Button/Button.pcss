@import "/src/styles/settings.pcss";

.button {
    box-sizing: border-box;
    display: inline-block;
    padding: 14px 30px;
    border: none;
    color: #fff;
    background-color: var(--green);
    font-family: var(--base-font-family);
    font-weight: var(--base-font-weight);
    font-size: 15px;
    line-height: 1.5;
    text-align: center;
    text-decoration: none;
    outline: none;
    user-select: none;
    overflow: hidden;
    text-overflow: ellipsis;
    border-radius: 0;
    cursor: pointer;
    transition: var(--tr_time);

    &:hover {
        background-color: var(--green-dark);
    }

    &:focus {
        outline: none;
        box-shadow: inset 0 0 0 1px var(--green-dark);
    }

    &:disabled,
    &_inactive {
        color: var(--gray_50);
        background-color: var(--gray_90);
        cursor: not-allowed;

        &:hover,
        &:focus {
            color: var(--gray_50);
            background-color: var(--gray_90);
            box-shadow: inset 0 0 0 1px var(--gray_90);
        }
    }

    &_inactive {
        cursor: pointer;
    }

    &_small {
        padding: 9px 19px;
    }

    &_wide {
        width: 100%;
    }
}

.button.button_black {
    
    background-color: var(--gray_15);

    &:hover {
        background-color: var(--gray_5);
    }

    &:focus {
        outline: none;
        box-shadow: inset 0 0 0 1px var(--gray_0);
    }
}

.button.button_gray {

    color: var(--black);
    background-color: var(--gray_90);

    &:hover {
        background-color: var(--gray_85);
    }

    &:focus {
        box-shadow: inset 0 0 0 1px var(--gray_85);
    }
}
.button.button_red {
    background-color: #d32f2f;
    color: #fff;

    &:hover {
        background-color: #b71c1c;
    }

    &:focus {
        outline: none;
        box-shadow: inset 0 0 0 1px #b71c1c;
    }
}

.button-loading {
    position: relative;
    cursor: not-allowed !important;
}

.loading {
    position: absolute;
    top: 50%;
    right: 50%;
    transform: translate(50%, -50%);

    font-weight: bold;
    display: inline-block;
    font-family: monospace;
    font-size: 15px;
    color: var(--white);
    width: 23px;
    height: 14px;
    background: radial-gradient(farthest-side, currentColor 90%, #0000),
        radial-gradient(farthest-side, currentColor 90%, #0000), radial-gradient(farthest-side, currentColor 90%, #0000);
    background-size: 6px 6px;
    background-repeat: no-repeat;
    animation: m 1.5s ease-out infinite;
}
@keyframes m {
    0% {
        background-position:
            calc(0 * 100% / 2) 0%,
            calc(1 * 100% / 2) 0%,
            calc(2 * 100% / 2) 0%;
        background-size:
            0px 0px,
            0px 0px,
            0px 0px;
    }
    16% {
        background-position:
            calc(0 * 100% / 2) 50%,
            calc(1 * 100% / 2) 0%,
            calc(2 * 100% / 2) 0%;
        background-size:
            5px 5px,
            0px 0px,
            0px 0px;
    }
    32% {
        background-position:
            calc(0 * 100% / 2) 100%,
            calc(1 * 100% / 2) 50%,
            calc(2 * 100% / 2) 0%;
        background-size:
            5px 5px,
            5px 5px,
            0px 0px;
    }
    48% {
        background-position:
            calc(0 * 100% / 2) 50%,
            calc(1 * 100% / 2) 100%,
            calc(2 * 100% / 2) 50%;
        background-size:
            5px 5px,
            5px 5px,
            5px 5px;
    }
    64% {
        background-position:
            calc(0 * 100% / 2) 0%,
            calc(1 * 100% / 2) 0%,
            calc(2 * 100% / 2) 100%;
        background-size:
            0px 0px,
            5px 5px,
            5px 5px;
    }
    80% {
        background-position:
            calc(0 * 100% / 2) 0%,
            calc(1 * 100% / 2) 0%,
            calc(2 * 100% / 2) 50%;
        background-size:
            0px 0px,
            0px 0px,
            5px 5px;
    }
    100% {
        background-position:
            calc(0 * 100% / 2) 0%,
            calc(1 * 100% / 2) 0%,
            calc(2 * 100% / 2) 0%;
        background-size:
            0px 0px,
            0px 0px,
            0px 0px;
    }
}

/* @keyframes l {
    to {
      clip-path: inset(0 -1ch 0 0)
    }
} */

.children-container {
    opacity: 0;
}
