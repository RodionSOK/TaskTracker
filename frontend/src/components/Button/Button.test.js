import React from "react";
import Button from "./Button.jsx";
import renderer from "react-test-renderer";

it("Button default", () => {
    const tree = renderer.create(<Button>button</Button>).toJSON();

    expect(tree).toMatchSnapshot();
});

it("Button disabled", () => {
    const tree = renderer.create(<Button disabled>button</Button>).toJSON();

    expect(tree).toMatchSnapshot();
});

it("Button small size ", () => {
    const tree = renderer.create(<Button size="small">button</Button>).toJSON();

    expect(tree).toMatchSnapshot();
});

it("Button with gray theme", () => {
    const tree = renderer.create(<Button theme="gray">button</Button>).toJSON();

    expect(tree).toMatchSnapshot();
});

it("Button with gray theme, small size", () => {
    const tree = renderer
        .create(
            <Button theme="gray" size="small">
                button
            </Button>,
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

it("Button inactive", () => {
    const tree = renderer.create(<Button inactive>button</Button>).toJSON();

    expect(tree).toMatchSnapshot();
});

it("Button full width", () => {
    const tree = renderer.create(<Button wide>button</Button>).toJSON();

    expect(tree).toMatchSnapshot();
});
