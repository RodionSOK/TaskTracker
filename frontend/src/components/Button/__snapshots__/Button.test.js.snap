// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default 1`] = `
<button
  className="button"
  disabled={undefined}
  onClick={[Function]}
>
  button
</button>
`;

exports[`Button disabled 1`] = `
<button
  className="button"
  disabled={true}
  onClick={[Function]}
>
  button
</button>
`;

exports[`Button full width 1`] = `
<button
  className="button button_wide"
  disabled={undefined}
  onClick={[Function]}
>
  button
</button>
`;

exports[`Button inactive 1`] = `
<button
  className="button button_inactive"
  disabled={undefined}
  onClick={[Function]}
>
  button
</button>
`;

exports[`Button small size  1`] = `
<button
  className="button button_small"
  disabled={undefined}
  onClick={[Function]}
>
  button
</button>
`;

exports[`Button with gray theme 1`] = `
<button
  className="button button_gray"
  disabled={undefined}
  onClick={[Function]}
>
  button
</button>
`;

exports[`Button with gray theme, small size 1`] = `
<button
  className="button button_small button_gray"
  disabled={undefined}
  onClick={[Function]}
>
  button
</button>
`;
